cmd/snap-repair/runner.go

```diff
@@ -185,3 +211,268 @@ func (run *Runner) Peek(brandID, repairID string) (headers map[string]interface{
 
 	return headers, nil
 }
+
+// deviceInfo captures information about the device.
+type deviceInfo struct {
+	Brand string `json:"brand"`
+	Model string `json:"model"`
+}
+
+// RepairStatus represents the possible statuses of a repair.
+type RepairStatus int
+
+const (
+	RetryStatus RepairStatus = iota
+	NotApplicableStatus
+	DoneStatus
+)
+
+// RepairState holds the current revision and status of a repair in a sequence of repairs.
+type RepairState struct {
+	Seq      int          `json:"seq"`
+	Revision int          `json:"revision"`
+	Status   RepairStatus `json:"status"`
+}
+
+// state holds the atomically updated control state of the runner with sequences of repairs and their states.
+type state struct {
+	Device    deviceInfo                `json:"device"`
+	Sequences map[string][]*RepairState `json:"sequences,omitempty"`
+}
+
+func (run *Runner) readState() error {
+	r, err := os.Open(dirs.SnapRepairStateFile)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	dec := json.NewDecoder(r)
+	return dec.Decode(&run.state)
+}
+
+func (run *Runner) initState() error {
+	if err := os.MkdirAll(dirs.SnapRepairDir, 0775); err != nil {
+		panic(fmt.Sprintf("cannot create repair state directory: %v", err))
+	}
+	// best-effort remove old
+	os.Remove(dirs.SnapRepairStateFile)
+	// TODO: init Device
+	run.SaveState()
+	return nil
+}
+
+// LoadState loads the repairs' state from disk, and (re)initializes it if it's missing or corrupted.
+func (run *Runner) LoadState() error {
+	err := run.readState()
+	if err == nil {
+		return nil
+	}
+	// error => initialize from scratch
+	// TODO: log error?
+	return run.initState()
+}
+
+// SaveState saves the repairs' state to disk.
+func (run *Runner) SaveState() {
+	m, err := json.Marshal(&run.state)
+	if err != nil {
+		panic(fmt.Sprintf("cannot marshal repair state: %v", err))
```

Comment:
That's more of a strange case for failure indeed, but I think we can still report it as a normal error, perhaps prefixing with "internal error: " to give a hint that we don't know why that happened. The harsh crash probably won't help I'm afraid.